### DISCUSSION TITLE
WHEP client recvonly instead of sendrecv direction?

### DISCUSSION THREAD

**Discussion by rgl** [bluenviron/mediamtx#2508] [Questions] [answered]
### Question

According to https://www.ietf.org/archive/id/draft-murillo-whep-02.html#section-4:

> The SDP offer SHOULD use the "recvonly" attribute and the SDP answer MUST use "sendonly" the attribute.

The only mention that I can about the SDP direction is at https://github.com/bluenviron/mediamtx/blob/v1.2.0/internal/core/webrtc_read_index.html#L162-L169:

```javascript
        const direction = "sendrecv";
        this.pc.addTransceiver("video", { direction });
        this.pc.addTransceiver("audio", { direction });
```

Shouldn't that be `recvonly` instead of `sendrecv`?

**Comment by aler9** [comment, marked as answer]
Hello, the reason why sendrecv is used in place of recvonly is that it allows H264 to work on Chrome and Firefox. it probably has to do with the codec preferences mechanism of the [WebRTC specification](https://www.w3.org/TR/webrtc/), that instructs peer to reject codecs specified by the counterpart in case of recvonly:

> The codec preferences of a [media description](https://www.w3.org/TR/webrtc/#dfn-media-description)'s [associated](https://www.w3.org/TR/webrtc/#dfn-associated) transceiver is said to be the value of the [RTCRtpTransceiver](https://www.w3.org/TR/webrtc/#dom-rtcrtptransceiver).[[[PreferredCodecs]]](https://www.w3.org/TR/webrtc/#dfn-preferredcodecs) with the following filtering applied (or said not to be set if [[[PreferredCodecs]]](https://www.w3.org/TR/webrtc/#dfn-preferredcodecs) is empty):
> 
> If the [direction](https://www.w3.org/TR/webrtc/#dom-rtcrtptransceiver-direction) is "[sendrecv](https://www.w3.org/TR/webrtc/#dom-rtcrtptransceiverdirection-sendrecv)", exclude any codecs not included in the intersection of [RTCRtpSender](https://www.w3.org/TR/webrtc/#dom-rtcrtpsender).[getCapabilities](https://www.w3.org/TR/webrtc/#dom-rtcrtpsender-getcapabilities)(kind).[codecs](https://www.w3.org/TR/webrtc/#dom-rtcrtpcapabilities-codecs) and [RTCRtpReceiver](https://www.w3.org/TR/webrtc/#dom-rtcrtpreceiver).[getCapabilities](https://www.w3.org/TR/webrtc/#dom-rtcrtpreceiver-getcapabilities)(kind).[codecs](https://www.w3.org/TR/webrtc/#dom-rtcrtpcapabilities-codecs).
> 
> If the [direction](https://www.w3.org/TR/webrtc/#dom-rtcrtptransceiver-direction) is "[sendonly](https://www.w3.org/TR/webrtc/#dom-rtcrtptransceiverdirection-sendonly)", exclude any codecs not included in [RTCRtpSender](https://www.w3.org/TR/webrtc/#dom-rtcrtpsender).[getCapabilities](https://www.w3.org/TR/webrtc/#dom-rtcrtpsender-getcapabilities)(kind).[codecs](https://www.w3.org/TR/webrtc/#dom-rtcrtpcapabilities-codecs).
> 
> If the [direction](https://www.w3.org/TR/webrtc/#dom-rtcrtptransceiver-direction) is "[recvonly](https://www.w3.org/TR/webrtc/#dom-rtcrtptransceiverdirection-recvonly)", exclude any codecs not included in [RTCRtpReceiver](https://www.w3.org/TR/webrtc/#dom-rtcrtpreceiver).[getCapabilities](https://www.w3.org/TR/webrtc/#dom-rtcrtpreceiver-getcapabilities)(kind).[codecs](https://www.w3.org/TR/webrtc/#dom-rtcrtpcapabilities-codecs).
> 
> The filtering MUST NOT change the order of the codec preferences.

Although H264 is present in `RTCRtpReceiver.getcapabilities()` of most browsers, so `recvonly` should be enough in theory, while it is not.

Anyway, further investigations are not worth the effort since changing a flag is enough to solve the issue.

[a comment by an automated account is not shown]
